hashintel/hash
# H-325: Improve how we define system types

## 🌟 What is the purpose of this PR?




This PR changes how we seed the system types in the Graph API. Previously we were initalizing them in a way that required a fresh datastor to ensure all types were seeded correctly, and did not account for migrating system types in an existing datastore.

This PR replaces the system type intitializers with migration files in `apps/hash-api/src/graph/migrate-ontology-types/migrations`, which for ensure the system types used by HASH are created, and ensures the external `blockProtocol` data types have been loaded. We will want to extend these migration files once we want to apply updates to existing system types, or create new ones.

In addition, this PR removes the shared `systemTypes` and `blockProtocolTypes` objects from `libs/@local/hash-isomorphic-utils/src/ontology-types.ts`, as these assume that all system types and block protocol types have a fixed version. These have been replaced with equivalent objects in `libs/@local/hash-isomorphic-utils/src/ontology-type-ids.ts`, which is a generated file containing the latest version of the types. This file can be generated via the `generate-ontology-type-ids` script, which requires the Graph API to be running.

## 🔗 Related links





- H-325

## 🔍 What does this change?




- adds a patch to the `@blockprtocol/graph` `codegen` function to allow for rewriting type IDs so that these are fetched from `localhost:3000` even when they actual type ID starts with `https://hash.ai/...`

## Pre-Merge Checklist 🚀

### 🚢 Has this modified a publishable library?




This PR:


### 📜 Does this require a change to the docs?




The changes in this PR:


### 🕸️ Does this require a change to the Turbo Graph?




The changes in this PR:


## ⚠️ Known issues



- we still have the linear system types defined via the initializers approach, and subsequently still have the `linearTypes` object in `libs/@local/hash-isomorphic-utils/src/ontology-types.ts`. When we redefine how these types are seed in upcoming integrations work we can remove this file and related initializer functions in `apps/hash-api/src/graph/util.ts`.

## 🐾 Next steps



## 🛡 What tests cover this?



-

## ❓ How to test this?



1. Re-create a locally running instance of HASH with a fresh database
2. Run the backend via `yarn dev:backend`, which will run the migration scripts and create the relevant system/blockProtocol types
3. Try out the generate ontology type IDs script via `yarn generate-ontology-type-ids`
4. Run the frontend via `yarn dev:frontend` and ensure the FE works as expected

## 📹 Demo
